cluster-autoscaler/cloudprovider/azure/azure_cache.go

```diff
@@ -353,6 +378,46 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
 	return m.autoscalingOptions[ref]
 }
 
+// HasInstance returns if a given instance exists in the azure cache
+func (m *azureCache) HasInstance(providerID string) (bool, error) {
+	vmsPoolSet := m.getVMsPoolSet()
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	resourceID, err := convertResourceGroupNameToLower(providerID)
+	if err != nil {
+		// Most likely an invalid resource id, we should return false
+		// most of these shouldn't make it here do to higher level
+		// validation in the HasInstance azure.cloudprovider function
+		return false, err
+	}
+
+	nodeGroupNameForNode := getVMSSNameFromProviderID(providerID)
+	inst := azureRef{Name: resourceID}
+	klog.V(4).Infof("HasInstance: Checking instance %s", inst.Name)
+	if m.unownedInstances[inst] || (m.vmType == vmTypeVMSS && len(vmsPoolSet) == 0 && !m.registeredAKSPoolNames.Has(nodeGroupNameForNode)) {
+		klog.V(4).Infof("HasInstance: Instance %s is unowned", inst.Name)
+		return false, cloudprovider.ErrNotImplemented
```

Comment:
I don't think the errors should be `cloudprovider.ErrNotImplemented`, but should be real error reflecting the reason.

Reply:
Have not yet look at all the logic here, but (see my comment elsewhere) `ErrNotImplemented` causes _silent_ fallback, which would likely be appropriate for unowned instances. Otherwise, in presence of non-autoscaled nodes, this would be spamming the logs all the time.

Reply:
@comtalyst  see silent fallback here https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/clusterstate/clusterstate.go#L1055